sklearn/ensemble/gradient_boosting.py

```diff
@@ -1419,11 +1466,26 @@ def fit(self, X, y, sample_weight=None, monitor=None):
             # init state
             self._init_state()
 
-            # fit initial model - FIXME make sample_weight optional
-            self.init_.fit(X, y, sample_weight)
+            # fit initial model and initialize raw predictions
+            if self.init == 'zero':
+                raw_predictions = np.zeros(shape=(X.shape[0], self.loss_.K),
+                                           dtype=np.float64)
+            else:
+                supports_sample_weight = has_fit_parameter(self.init_,
+                                                           "sample_weight")
+                if sample_weight_is_none:
+                    self.init_.fit(X, y)
+                else:
+                    if not supports_sample_weight:
```

Comment:
Nit: We do not need to define `supports_sample_weight` and have it in the if statement here:

```python
if not has_fit_parameter(self.init_, "sample_weight")
```

Reply:
@jnothman argued against using `has_fit_parameter`, see [this comment](https://github.com/scikit-learn/scikit-learn/pull/12436#discussion_r228739257). That's why I used a try except instead:
```python
try:
    self.init_.fit(X, y, sample_weight=sample_weight)
except TypeError:
    if sample_weight_is_none:
        self.init_.fit(X, y)
    else:
        raise ValueError(
            "The initial estimator {} does not support sample "
            "weights.".format(self.init_.__class__.__name__))
```